### PR TITLE
bugfix: freed all connections when cleaning up socket pool

### PR DESCRIPTION
**Pull request by spacewander**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_socket_tcp.c [hunk 1]
```diff
@@ -5035,6 +5033,17 @@ ngx_http_lua_socket_shutdown_pool(lua_State *L)
     }
 
     spool->active_connections = 0;
+}
+
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: Style: needs 2 blank lines to separate C function definitions.

File: src/ngx_http_lua_socket_tcp.c [hunk 2]
```diff
@@ -5463,21 +5469,10 @@ ngx_http_lua_cleanup_conn_pools(lua_State *L)
...
-            ngx_http_lua_socket_tcp_close_connection(c);
-
-            ngx_queue_remove(q);
-
-            ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
-                           "lua tcp socket keepalive: free connection pool "
-                           "for \"%s\"", spool->key);
-        }
-
+        ngx_http_lua_socket_shutdown_pool_helper(spool);
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: Better add blank lines around this line to make this line stand out.

File: src/ngx_http_lua_socket_tcp.c [hunk 3]
```diff
@@ -122,7 +120,13 @@ static ngx_int_t ngx_http_lua_socket_tcp_resume_helper(ngx_http_request_t *r,
     int socket_op);
 static void ngx_http_lua_tcp_resolve_cleanup(void *data);
 static void ngx_http_lua_coctx_cleanup(void *data);
+
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: Better remove this blank line.

File: src/ngx_http_lua_socket_tcp.c [hunk 4]
```diff
@@ -122,7 +120,13 @@ static ngx_int_t ngx_http_lua_socket_tcp_resume_helper(ngx_http_request_t *r,
...
+static void ngx_http_lua_socket_free_pool(ngx_log_t *log,
+    ngx_http_lua_socket_pool_t *spool);
 static int ngx_http_lua_socket_shutdown_pool(lua_State *L);
+static void ngx_http_lua_socket_shutdown_pool_helper(
+    ngx_http_lua_socket_pool_t *spool);
+
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: Ditto.

**Comment by spacewander** [comment]
@agentzh 
The code is updated.

**Comment by agentzh** [comment]
@spacewander I wonder if it's possible to add some test cases to cover this fix? This way we cannot make the same mistake in the future again ;)

**Comment by spacewander** [comment]
@agentzh 
Have not got an idea yet...
The socket pool clean up is called only when ngx_lua destroy the Lua VM, but GC will free the connections too. So the connections are freed eventually whether we have freed them in cleaning up or not.

**Comment by agentzh** [comment]
@spacewander OK, understood. Merged with the following minor extra patch:

```diff
diff --git a/src/ngx_http_lua_socket_tcp.c b/src/ngx_http_lua_socket_tcp.c
index c3be3999..68691122 100644
--- a/src/ngx_http_lua_socket_tcp.c
+++ b/src/ngx_http_lua_socket_tcp.c
@@ -5040,6 +5040,7 @@ ngx_http_lua_socket_shutdown_pool(lua_State *L)
     ngx_http_lua_socket_pool_t          *spool;

     spool = lua_touserdata(L, 1);
+
     if (spool != NULL) {
         ngx_http_lua_socket_shutdown_pool_helper(spool);
     }
@@ -5469,11 +5470,14 @@ ngx_http_lua_cleanup_conn_pools(lua_State *L)
         /* tb key val */
         spool = lua_touserdata(L, -1);

-        ngx_http_lua_socket_shutdown_pool_helper(spool);
+        if (spool != NULL) {
+            ngx_log_debug2(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                           "lua tcp socket keepalive: free connection pool %p "
+                           "for \"%s\"", spool, spool->key);
+
+            ngx_http_lua_socket_shutdown_pool_helper(spool);
+        }

-        ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
-                       "lua tcp socket keepalive: free connection pool "
-                       "for \"%s\"", spool->key);
         lua_pop(L, 1);
     }
```

Thanks!

**Pull request closed without merging**